bricklayers.py

```diff
@@ -1,143 +1,282 @@
-                    else: 
-                        new_e_value = e_value * extrusion_multiplier
-                        logging.info(f"Multiplying E value by extrusionMultiplier")
-                        line = re.sub(r'E[-\d.]+', f'E{new_e_value:.5f}', line).strip()
-                        line += f" ; Adjusted E for extrusionMultiplier, block #{perimeter_block_count}\n"
-						
-        elif perimeter_type == "internal" and line.startswith("G1") and "X" in line and "Y" in line and "F" in line:  # End of perimeter block
-            inside_perimeter_block = False
-
-        modified_lines.append(line)
-
-    # Overwrite the input file with the modified G-code
-    with open(input_file, 'w') as outfile:
-        outfile.writelines(modified_lines)
-
-    logging.info("G-code processing completed")
-    logging.info(f"Log file saved at {log_file_path}")
-
-# Main execution
-if __name__ == "__main__":
-    parser = argparse.ArgumentParser(description="Post-process G-code for Z-shifting and extrusion adjustments.")
-    parser.add_argument("input_file", help="Path to the input G-code file")
-    parser.add_argument("-layerHeight", type=float, default=0.2, help="Layer height in mm (default: 0.2mm)")
-    parser.add_argument("-extrusionMultiplier", type=float, default=1, help="Extrusion multiplier for first layer (default: 1.5x)")
-    args = parser.parse_args()
-
-    process_gcode(
-        input_file=args.input_file,
-        layer_height=args.layerHeight,
-        extrusion_multiplier=args.extrusionMultiplier,
-    )
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see .
+#
+# Copyright (c) [2025] [Roman Tenger]
+
+import re
+import sys
+import logging
+import os
+import argparse
+from tempfile import NamedTemporaryFile
+import io
+from pathlib import Path
+
+class GCodeProcessor:
+    def __init__(self, layer_height, extrusion_multiplier):
+        self.layer_height = layer_height
+        self.extrusion_multiplier = extrusion_multiplier
+        self.current_layer = 0
+        self.current_z = 0.0
+        self.perimeter_type = None
+        self.perimeter_block_count = 0
+        self.inside_perimeter_block = False
+        self.z_shift = layer_height * 0.5
+        self.total_layers = 0
+        self.buffer_size = 1024 * 1024  # 1MB buffer size
+        self.printer_type = None
+        self.in_object = False
+        self.shifted_blocks = 0
+        self.perimeter_found = False
```

Comment:
IMO it's much simpler/clearer to just explicitly pass these around where they're needed, vs dropping all the state in this class